### ISSUE TITLE
[Feature Request] Easing gradients

### ISSUE THREAD

**Issue by zslabs** [styled-components/polished#441]
👋 Hi all! I've been a big fan of polished; especially coming from the Sass/PostCSS world. I'm working on a project with Emotion, and am missing a few common PostCSS plugins I used to use; one of them being https://github.com/larsenwork/postcss-easing-gradients

Any thoughts on adding something like that to polished? Thanks!

**Comment by JacobCooley** [comment]
I was curious about this too!

**Comment by bhough** [comment]
We will be adding this in v4.

**Comment by tomekbuszewski** [comment]
> We will be adding this in v4.

Any news on this? I can't find this documented anywhere.

**Comment by bhough** [comment]
@tomekbuszewski We have not released v4 yet. But we are looking at possibly pulling it forward in v3.5 that will be out sometime in the next 4-6 weeks.

**Comment by ricokahler** [comment]
Hey all, around 2 weeks ago, I was emailing @bhough about potentially taking this feature on. After digging into it I have a few suggestions on the API and implementation but it requires some relatively large changes and I want to pitch it before going all in.

## API

I think the best place to integrate this feature is to add it to the existing `linearGradient` and `radialGradient` functions.

Looking at CSSWG Drafts repo, it looks like the syntax [they're converging on (back in 2017 lol)](https://github.com/w3c/csswg-drafts/issues/1332#issuecomment-299990698), is `<color-stop>, <timing-function>, <color-stop>, (repeat)`. Given that, I think an API that makes sense is to mimic that decision and allow timing functions to be provided in the current `colorStops` arg.

E.g.

```js
linearGradient({
  colorStops: [
    '#00FFFF 0%',
    'ease-in-out',
    'rgba(0, 0, 255, 0) 50%',
    'ease-in-out',
    '#0000FF 95%'
   ],
  toDirection: 'to top right',
});
```

If a timing function isn't provided, we default to linear.

## Implementation

We can already interpolate linearly via `mix` so I think it makes sense to use that, but easing `mix` requires some manual math. In short, I think it makes sense to replace [the strings](https://github.com/styled-components/polished/blob/cd95bb3f1ab4b8c6590a69dda8b3d841f3894ac7/src/easings/easeIn.js#L5-L12) of the cubic bezier easings and break them up into their raw numbers, that way we can plug those numbers into the [cubic bezier function in it's explicit form to interpolate with easing.](https://en.wikipedia.org/wiki/B%C3%A9zier_curve#Cubic_B.C3.A9zier_curves)

For example:

```js
const easeInSine = [0.470,  0.000, 0.745, 0.715];

function createTimingFn([p0, p1, p2, p3]) {
  return t => (
    (1 - t) ** 3 * p0
    + 3 * (1 - t) ** 2 * t * p1
    + 3 * (1 - t) * t ** 2 * p2
    + t ** 3 * p3
  );
}

const timing = createTimingFn(easeInSine)

const stops = [];

for (let i = 0; i < 15; i += 1) {
  stops.push(mix(timing(i / 14), 'black', 'transparent'));
}
```

The actual implementation of eased gradients still require parsing the color percentage stops and doing more math but the gist of the implementation is using the above technique to interpolate between each color stop.

## Possibility of adding a `getScale` function?

An interesting side-effect of the approach above is that it requires us to implement a color scale function internally to make the eased gradients work. This would be similar to chroma's [`scale` function](https://gka.github.io/chroma.js/#chroma-scale).

Since we have to build this function anyway to power eased gradients, why not expose this scale function externally? This function would take in same type of colors stops as the linear gradient function and return another function that takes in a number between 0 and 1 and returns the color at the point in the scale.

---

To summarize:

1. Proposed API extends the current one, taking in timing functions between color stops
2. We can implement the eased gradients with `mix` but this requires implementing a cubic bezier function and exposing all of the easing numbers that are currently in strings.
3. As a bonus, we can easily expose a `getScale` function similar to chroma's `scale` function.

So all of this works but it's a larger change. [The related PostCSS plugin](https://github.com/larsenwork/postcss-easing-gradients) gets away with it because they only support a smaller syntax subset and the code they write doesn't ship to browsers. I think the proposed API makes sense but I'm not 100% convinced these changes are worth it.

Let me know what you think!

**Comment by zslabs** [comment]
@ricokahler Thank you so much for writing this up in such detail! When I was originally looking into this feature myself, the `scale` function you mentioned was also a stumbling block for myself with the weight behind the library. I certainly wouldn't want to bloat the size of this library for this one use-case, unless Chroma was seen as a benefit that could replace other types of color manipulation across the board.

**Comment by ricokahler** [comment]
@zslabs if we reuse `mix` and migrate the current easing strings to number arrays, then adding something like `getScale` might not have that a big impact on bundle size (thought I'm still concerned with the amount of changes required to get it done).

What might add more bloat bundle-size wise would be parsing the linearGradient color stop arrays. Right now we get away with allowing whatever the browser supports because we just [pass it straight to the browser 🤷‍♀️](https://github.com/styled-components/polished/blob/b0a95716173c916dcb57d904001ce9aebf394e67/src/mixins/linearGradient.js#L47-L49). However, if you look at what [the CSS spec supports, it's non-trivial to parse](https://developer.mozilla.org/en-US/docs/Web/CSS/linear-gradient#Customizing_Gradients) if we want to normalize and do math on this array.

If we implement the easing ourselves then we need a normalized version of the colors stops because we need know what value to use when in-between colors. Parsing those strings and the edge cases would probably be too much code so I thinking about this more, we should probably either stick to a simpler syntax (e.g. maybe the colors stops are always an `Array<{color: string, percentage?: string, timingFunction?: string}>` instead of pure CSS strings) or we can abandon my original idea of re-using the linear gradient API and just expose a simpler function that can do eased gradients.

Maybe we can allow the `mix` function to take in a timing function and then `mix` can be what users use to implement eased gradients. That's a pretty low-level solution so we'd need some good docs to go along with that.

---

My current vote is to change the `colorStops` arg of `linearGradient`/`radialGradient` array type because it keeps the visibility of the feature. My intuition says this feature would add maybe like 500Bish? of gzipped code if we reuse + migrate correctly but the only way to find out for real is to implement it.

Some questions for @bhough:

1. Are you okay with the amount of changes required to implement this feature? (e.g. migrating the easing strings to number arrays, adding an internal timing function tool at a minimum)
2. What is your bundle size budget for this feature?
3. What kinds of external API changes are you comfortable with? If we change the `colorStops` arg type in v4, would that be okay?

**Comment by bhough** [comment]
@ricokahler Generally speaking we try to get users a full version to transition their APIs. We tend to be willing to trade-off a slightly larger bundle size in the interim to support both APIs until we fully deprecate.

- If I'm eye-balling it correctly, I believe the change to number based easing values could actually reduce the size, all things being equal.
- We try to mimick CSS wherever possible, so agree the implementation should be the same.
- I support adding a scale function and exposing it, but I would keep it limited to base implementation for now and hold off on the extra weight of the extra methods on it (i.e. `.domain)
- We don't worry too much but overall bundle budgets. Because we expect users to be tree-shaking and the mono-repo/submodules we are introducing in 4.0 will further reduce that concern.
- The shared API signatures between eased and standard gradients is tricky. We have gotten ourselves in trouble in the past trying to over support different signatures to ease use. Tackling Flow/TS support can get especially challenging.
- .5kb is definitely a large hit for a more niche feature, especially if folks not using it are forced to bring the bundle size onboard. That being said, my current opinion is if we can't get it to fit directly into linear/gradient in an economical fashion, we can bring it in as standalone module(s).
- One other option, given that we already provide a less CSS-like API for these, is to add an easing argument that also takes an array of easing functions. Then we only need to check that they are valid values and that the array length equals the colorStop length. 

I see no reason not to start diving into the code and see where we end up.

**Comment by ricokahler** [comment]
Hey Brian,

I know I kind of gave up on the eased gradient feature but I just saw Lea Verou un-private this repo: https://github.com/LeaVerou/color.js <https://github.com/LeaVerou/color.js>

It could greatly help the development of LCH support. If there’s anyone to trust to get LCH right, it’s her lol.

Hope this is useful and sorry for giving up there. Best of luck with polished!

**Comment by bhough** [comment]
After several attempts to implement, we've decided the code needed to do so is to heavy for the niche functionality it brings. We are open to revisiting this if a lighterweight way to implement it can be found.

**Comment by zslabs** [comment]
@bhough Appreciate you vetting this! Glad to hear the team decided the weight/complexity involved wasn't worth it. Thanks again!
